packages/grafana-ui/src/components/Splitter/Splitter.tsx

```diff
@@ -0,0 +1,482 @@
+        splitterRef.current!.ariaValueNow = '100';
+        return;
+      }
+
+      if (
+        !(
+          (direction === 'column' && VERTICAL_KEYS.has(e.key)) ||
+          (direction === 'row' && HORIZONTAL_KEYS.has(e.key))
+        ) ||
+        pressedKeys.current.has(e.key)
+      ) {
+        return;
+      }
+
+      savedPos.current = undefined;
+      e.preventDefault();
+      e.stopPropagation();
+      primarySizeRef.current = firstPaneRef.current!.getBoundingClientRect()[measurementProp];
+      containerSize.current = containerRef.current!.getBoundingClientRect()[measurementProp];
+      firstPaneMeasurements.current = measureElement(firstPaneRef.current!);
+      const newKey = !pressedKeys.current.has(e.key);
+
+      if (newKey) {
+        const initiateAnimationLoop = pressedKeys.current.size === 0;
+        pressedKeys.current.add(e.key);
+
+        if (initiateAnimationLoop) {
+          window.requestAnimationFrame(handlePressedKeys);
+        }
+      }
+    },
+    [direction, handlePressedKeys, handleSize, maxDimProp, measurementProp, minDimProp]
+  );
+
+  const onKeyUp = useCallback(
+    (e: React.KeyboardEvent<HTMLDivElement>) => {
+      if (
+        (direction === 'row' && !HORIZONTAL_KEYS.has(e.key)) ||
+        (direction === 'column' && !VERTICAL_KEYS.has(e.key))
+      ) {
+        return;
+      }
+
+      pressedKeys.current.delete(e.key);
+      onSizeChange?.(parseFloat(firstPaneRef.current!.style.flexGrow));
+    },
+    [direction, onSizeChange]
+  );
+
+  const onDoubleClick = useCallback(() => {
+    firstPaneRef.current!.style.flexGrow = '0.5';
+    secondPaneRef.current!.style.flexGrow = '0.5';
+    const dim = measureElement(firstPaneRef.current!);
+    firstPaneMeasurements.current = dim;
+    primarySizeRef.current = firstPaneRef.current!.getBoundingClientRect()[measurementProp];
+    splitterRef.current!.ariaValueNow = `${((primarySizeRef.current - dim[minDimProp]) / (dim[maxDimProp] - dim[minDimProp])) * 100}`;
+  }, [maxDimProp, measurementProp, minDimProp]);
+
+  const onBlur = useCallback(() => {
+    // If focus is lost while keys are held, stop changing panel sizes
+    if (pressedKeys.current.size > 0) {
+      pressedKeys.current.clear();
+      dragStart.current = null;
+      onSizeChange?.(parseFloat(firstPaneRef.current!.style.flexGrow));
+    }
+  }, [onSizeChange]);
+
+  return {
+    kids,
+    containerRef,
+    firstPaneRef,
+    minDimProp,
+    splitterRef,
+    measurementProp,
+    handleSize,
+    onPointerUp,
+    onPointerDown,
+    onPointerMove,
+    onKeyDown,
+    onKeyUp,
+    onDoubleClick,
+    onBlur,
+    secondPaneRef,
+  };
+}
+
+type MeasureR<T> = T extends HTMLElement
+  ? { minWidth: number; maxWidth: number; minHeight: number; maxHeight: number }
+  : T extends null
+    ? undefined
+    : never;
+type HTMLElementOrNull = HTMLElement | null;
+
+function measureElement<T extends HTMLElementOrNull>(ref: T): MeasureR<T> {
```

Comment:
i think we can simplify our types a lot here if we restrict this to only allow an element to be passed (and not `null`). then our return type is always just
```tsx
{ minWidth, maxWidth, minHeight, maxHeight }
```

and we just gate every use of `measureElement` behind an if statement, like
```tsx
if (firstPaneRef.current) {
  const dim = measureElement(firstPaneRef.current);
  ...
}
```

this will also allow us to remove all(?) of the `!` operators. maybe slightly more lines, but imo more readable and type-safe 😎

Reply:
good suggestion! updated, had to add a lot if statements though :)